sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts

```diff
@@ -0,0 +1,132 @@
+const delayWhenNoSessionsAvailableMs = 5 * 1000;
+
+function processInitialize(sessionId: string) {
+  console.log(`[${sessionId}] will start processing...`);
+}
+
+function processMessage(msg: ReceivedMessageWithLock) {
+  console.log(`[${msg.sessionId}] received message with body ${msg.body}`);
+}
+
+function processError(err: Error, sessionId: string) {
+  console.log(`[${sessionId}] had error ${err.message}`);
+}
+
+function processClose(reason: "error" | "idle_timeout", sessionId: string) {
+  if (reason === "error") {
+    console.log(`[${sessionId}] was closed because of error`);
+  } else if (reason === "idle_timeout") {
+    console.log(`[${sessionId}] was closed - no more messages within idle timeout`);
+  }
+}
+
+function _createIdleTimer(
+  timeoutMs: number
+): { refresh(): void; expirationPromise: Promise<void> } {
+  let resolveExpirationPromise: () => void;
+
+  const expirationPromise = new Promise<void>((res, _rej) => {
+    resolveExpirationPromise = res;
+  });
+
+  let timer = setTimeout(() => resolveExpirationPromise(), timeoutMs);
+
+  return {
+    refresh() {
+      clearTimeout(timer);
+      timer = setTimeout(() => resolveExpirationPromise(), timeoutMs);
+    },
+    expirationPromise
+  };
+}
+
+async function _processNextSession(
+  serviceBusClient: ServiceBusClient,
+  abortSignal: AbortSignalLike
+): Promise<void> {
+  let sessionReceiver: SessionReceiver<ReceivedMessageWithLock>;
+
+  try {
+    sessionReceiver = await serviceBusClient.createSessionReceiver(queueName, "peekLock", {
+      autoRenewLockDurationInMs: sessionIdleTimeoutMs
+    });
+  } catch (err) {
+    if ((err as MessagingError).code === "no-sessions-available") {
+      console.log(`INFO: no available sessions, sleeping for ${delayWhenNoSessionsAvailableMs}`);
+      await delay(delayWhenNoSessionsAvailableMs);
+      return;
+    }
+
+    throw err;
+  }
+
+  await processInitialize(sessionReceiver.sessionId);
+
+  const idleTimer = _createIdleTimer(sessionIdleTimeoutMs);
+  let threwError = false;
+
+  sessionReceiver.subscribe(
+    {
+      async processMessage(msg) {
+        idleTimer.refresh();
+        await processMessage(msg);
+      },
+      async processError(err) {
+        threwError = true;
+        await processError(err, sessionReceiver.sessionId);
+      }
+    },
+    {
+      abortSignal
+    }
+  );
+
+  await idleTimer.expirationPromise;
+
+  try {
+    await sessionReceiver.close();
+  } catch (err) {
+    await processError(err, sessionReceiver.sessionId);
+  }
+
+  await processClose(threwError ? "error" : "idle_timeout", sessionReceiver.sessionId);
+}
+
+async function _roundRobinThroughAvailableSessions(abortSignal: AbortSignalLike): Promise<void> {
+  const serviceBusClient = new ServiceBusClient(serviceBusConnectionString);
+
+  for (let i = 0; i < maxSessionsToProcessSimultaneously; ++i) {
+    (async () => {
+      while (!abortSignal.aborted) {
+        await _processNextSession(serviceBusClient, abortSignal);
```

Comment:
If `_processNextSession` throws an error, what happens to other concurrent sessions? Would they keep going or should we abort them?

Reply:
There was one spot where we propagated an error and actually threw (when we would createSessionReceiver). I changed this to be consistent and just funnel the error through the user's processError() function instead.

I think we're okay with the simple "just keep going forever" model. I moved the abortController definition up a bit higher and mentioned (in a comment) that you can use `abortController.abort` to terminate the session manager.